### PR TITLE
zstd: Add delta encoding support

### PR DESCRIPTION
**Pull request by greatroar**
This adds support for delta encoding to compress/zstd, compatible with the [--patch-from](https://github.com/facebook/zstd/wiki/Zstandard-as-a-patching-engine) option that was introduced in zstd reference v1.4.5.

The example is a bit simplistic for now.

### CODE REVIEW

File: zstd/example_test.go [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+	}
+
+	// Ordinary compression, for reference.
+	w, _ = zstd.NewWriter(nil, bestLevel)
+	z := w.EncodeAll(target, nil)
+	fmt.Printf("compression only: %4d bytes\n", len(z))
+
+	// Output:
+	// delta encoding:     44 bytes
+	// compression only:  111 bytes
```
Comment (klauspost) [inline zstd/example_test.go:46]: This will likely change. Tests should not rely on exact compression numbers.

File: zstd/encoder_options.go [hunk 2]
```diff
@@ -316,3 +322,14 @@ func WithEncoderDict(dict []byte) EOption {
...
 	}
 }
+
+// WithEncoderDictRaw registers a dictionary that may be used by the encoder.
+//
+// The slice content may contain arbitrary data. It will be used as an initial
+// history.
+func WithEncoderDictRaw(id uint32, content []byte) EOption {
+	return func(o *encoderOptions) error {
+		o.dict = &dict{id: id, content: content}
```
Comment (klauspost) [inline zstd/encoder_options.go:332]: offsets must be initialized to `[3]uint32{1, 4, 8}`.

**Review by klauspost — COMMENTED** [review]
Looks good! Some minor things.

File: zstd/encoder_options.go (hunk 2 above)
Comment (greatroar) [reply zstd/encoder_options.go]: Done. Also put in a check for dictionaries that are too large for the reference implementation.

File: zstd/example_test.go (hunk 1 above)
Comment (greatroar) [reply zstd/example_test.go]: I've put in a looser check.

**Review by klauspost — APPROVED** [review]
LGTM

File: zstd/decoder_options.go [hunk 3]
```diff
@@ -99,6 +105,15 @@ func WithDecoderDicts(dicts ...[]byte) DOption {
 	}
 }
 
+// WithEncoderDictRaw registers a dictionary that may be used by the decoder.
+// The slice content can be arbitrary data.
+func WithDecoderDictRaw(id uint32, content []byte) DOption {
+	return func(o *decoderOptions) error {
+		o.dicts = append(o.dicts, dict{id: id, content: content})
```
Comment (klauspost) [inline zstd/decoder_options.go:112]: Please add the same as you did to the encoder option.

**Review by klauspost — CHANGES_REQUESTED** [review]

File: zstd/decoder_options.go (hunk 3 above)
Comment (greatroar) [reply zstd/decoder_options.go]: Done.

**Review by klauspost — APPROVED** [review]

**Comment by klauspost** [comment]
Considered adding a note that the most common data should be at the *end* of the dictionary, since offsets will be smaller and it will compress better. Couldn't really phase it in an easy to understand way, so that can always be added later.

**Pull request merged**